### ISSUE TITLE
`withTimeout` with a long duration keeps the `DefaultExecutor` thread alive for much longer than necessary

### ISSUE THREAD

**Issue by chaoren** [Kotlin/kotlinx.coroutines#4063]
**Describe the bug**

What happened?

After `withTimeout`, the `DefaultExecutor` thread will frequently stay around for the entire length of the timeout specified even if the actual `withTimeout` block finishes immediately. This seems to be ignoring the "keep alive" duration configured for `DefaultExecutor`:

https://github.com/Kotlin/kotlinx.coroutines/blob/60d2fe8471ce4d289d526f04b51389b06b1c064f/kotlinx-coroutines-core/jvm/src/DefaultExecutor.kt#L32-L39

What should have happened instead?

The `DefaultExecutor` thread should only stay alive for at most one second after becoming idle according to  `DEFAULT_KEEP_ALIVE_MS`.

**Provide a Reproducer**

Run this a bunch of times:
```kotlin
fun main() {
    runBlocking { withTimeout(10.seconds) {} }
    val time = measureTime {
        arrayOfNulls<Thread>(Thread.activeCount())
            .apply(Thread::enumerate)
            .find { it?.name == "kotlinx.coroutines.DefaultExecutor" }!!
            .join()
    }
    println(time)
}
```
Sometimes the thread takes about one second to join as expected, but more often than not, it'll take 11 seconds instead. Seems to be the duration given to `withTimeout` plus one whenever that happens.

* If possible, please provide a small self-contained project (or even just a single file) where the issue reproduces.
* If you can't pinpoint the issue, please provide at least *some* project where this reproduces, for example, your production one. If you are not ready to show the project publicly, we are open to discussing the details privately.
* If you really can't provide any code, please do still open an issue. This may prompt other people to chime in with their reproducers.

**Comment by dkhalanskyjb** [comment]
Yes, that's correct, once `DefaultExecutor` sees that it's supposed to run some code eventually, it will just go to sleep and won't wake up until either the time comes to run that code, or some new event arrives. `KEEP_ALIVE_MS` is not part of the public contract and isn't advertised anywhere, so it's just a part of our internal implementation, not something you can rely on.

If you have specific requirements for when `DefaultExecutor` should finish, please share them. Without specific use cases, we'd rather not wake `DefaultExecutor` up each time a task is *removed* from its queue: let it sleep in peace.

Also, if you *are* okay with relying on the internal implementation details, you can `interrupt` the `DefaultExecutor` so that it recognizes that it doesn't have any more work to do:
```kotlin
runBlocking { withTimeout(10.seconds) {} }
val time = measureTime {
    arrayOfNulls<Thread>(Thread.activeCount())
        .apply(Thread::enumerate)
        .find { it?.name == "kotlinx.coroutines.DefaultExecutor" }!!
        .apply {
            interrupt()
            join()
        }
}
println(time)
```

**Comment by chaoren** [comment]
This came up because https://github.com/Kotlin/kotlinx.coroutines/commit/9a98eabed91c71cb2da9b0fe061139632e08990b changed timeout of `runTest` to 60 seconds, which caused the `DefaultExecutor` thread to far outlive the lifetime of each test case and it was caught by a leaked thread detector that some of our tests used. That thread should obviously have been exempt from the detector anyway so that wasn't a real problem, but I thought that maybe this behavior was unintentional after seeing that the thread was only supposed to live for one second.

> we'd rather not wake `DefaultExecutor` up each time a task is *removed* from its queue: let it sleep in peace.

Fair enough. I'll just consider this an implementation quirk then. No real problems here.
